### PR TITLE
[train] Sort workers by node ID rather than by node IP

### PR DESCRIPTION
**Pull request by KyleKoon**
## Why are these changes needed?

We should not sort workers by IP, since it may be the case that multiple nodes are placed on the same IP host. If multiple nodes are collocated, the local rank and node rank of the workers will be incorrect. We should instead use a node's unique ID to map to a set of workers.

Example:
``` 
Worker 0: node 0, ip 0
Worker 1: node 0, ip 0
Worker 2: node 1, ip 0
Worker 3: node 1, ip 0
```
Current Behavior:
```
worker 0: local_rank 0, node_rank 0
worker 1: local_rank 1, node_rank 0
worker 2: local_rank 2, node_rank 0
worker 3: local_rank 3, node_rank 0
```
Expected Behavior:
```
worker 0: local_rank 0, node_rank 0
worker 1: local_rank 1, node_rank 0
worker 2: local_rank 0, node_rank 1
worker 3: local_rank 1, node_rank 1
```

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by woshiyyya** [comment]
In general it looks good. Can we add some tests to reproduce the failure case we want to resolve (multiple workers nodes shares the  same IP)?

You can refer to this example to construct virtual clusters: https://github.com/ray-project/ray/blob/450df6017c354726d1dec114c03fe11de1348590/python/ray/train/tests/conftest.py#L45

File: python/ray/train/tests/test_backend.py [hunk 1]
```diff
@@ -178,6 +191,21 @@ def train_func():
         assert list(e.finish_training()) == [2, 2, 1]
 
 
+def test_local_world_size_with_same_ip_nodes(ray_2_node_2_cpu):
+    config = TestConfig()
+    with patch.object(
+        WorkerGroup, "add_workers", mock_add_workers_to_nodes_with_same_ip
```
Comment (woshiyyya) [inline python/ray/train/tests/test_backend.py:197]: The 2 virtual nodes will have the same node_ip. Can we also have a test without patching the `add_worker` method?

**Review by woshiyyya — APPROVED** [review]
It looks good to me. Left one comment

**Review by matthewdeng — APPROVED** [review]
Thanks!

**Pull request merged**